<a href="https://colab.research.google.com/github/alik604/Notebooks/blob/master/Data%20Science-Datasets/pima%20indians%20diabetes%20dataset/Diabetes_Prediction_cleaned_up_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

https://en.wikipedia.org/wiki/Decision_tree_learning#Decision_tree_types
https://en.wikipedia.org/wiki/Decision_tree_learning#Metrics

https://www.datacamp.com/community/tutorials/decision-tree-classification-python

images 
https://scikit-learn.org/stable/modules/tree.html


### regressor vs classifier: 
https://math.stackexchange.com/a/183001


https://www.kaggle.com/pouryaayria/a-complete-ml-pipeline-tutorial-acu-86

https://www.kaggle.com/shrutimechlearn/step-by-step-diabetes-classification-knn-detailed

https://www.kaggle.com/ash316/ml-from-scratch-part-2

In [59]:
import scipy 
import numpy as np
import pandas as pd 
from matplotlib import pyplot as plt

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split # Import train_test_split function

from sklearn.decomposition import PCA

from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier,export_graphviz # Import Decision Tree Classifier
from sklearn.ensemble import RandomForestClassifier,VotingClassifier, GradientBoostingClassifier
from sklearn.svm import SVC,SVR,LinearSVR,LinearSVC
from sklearn.naive_bayes import GaussianNB,BernoulliNB,BaseDiscreteNB,ComplementNB,MultinomialNB
from sklearn.ensemble import ExtraTreesClassifier

from sklearn.metrics import *

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [2]:
col_names = ['pregnant', 'glucose', 'bp', 'skin', 'insulin', 'bmi', 'pedigree', 'age', 'label']

pima = pd.read_csv('https://raw.githubusercontent.com/alik604/Notebooks/master/Data%20Science-Datasets/pima%20indians%20diabetes%20dataset/diabetes-cleanData.csv' ,header=0,error_bad_lines=False)

# keeping our fetures from above 
feature_cols = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age']
X = pima[feature_cols] # Features
y = pima['Outcome'] # Target variable

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1) # 70% training and 30% test

pima.head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            9      134             74             33       60  25.9   
1           10      101             76             48      180  32.9   
2            7      142             60             33      190  28.8   
3            6      129             90              7      326  19.6   
4            8      181             68             36      495  30.1   
5            1      189             60             23      846  30.1   
6            5      144             82             26      285  32.0   
7            8      176             90             34      300  33.7   
8           12      140             82             43      325  39.2   
9            0      173             78             32      265  46.5   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.460   81        0  
1                     0.171   63        0  
2                     0.687   61        0  
3                     0.582   60        0  
4                     0.615   60        1  
5                     0.398   59        1  
6                     0.452   58        1  
7                     0.467   58        1  
8                     0.528   58        1  
9                     1.159   58        0

In [3]:
pima.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   389.000000  389.000000     389.000000     389.000000  389.000000   
mean      3.208226  122.547558      70.642674      29.118252  156.210797   
std       3.041862   30.883067      12.538532      10.536701  119.249681   
min       0.000000   56.000000      24.000000       7.000000   14.000000   
25%       1.000000   99.000000      62.000000      21.000000   76.000000   
50%       2.000000  119.000000      70.000000      29.000000  126.000000   
75%       5.000000  143.000000      78.000000      37.000000  190.000000   
max      13.000000  198.000000     110.000000      63.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  389.000000                389.000000  389.000000  389.000000  
mean    33.046787                  0.523527   30.753213    0.326478  
std      7.038295                  0.345944   10.159084    0.469528  
min     18.200000                  0.085000   21.000000    0.000000  
25%     28.400000                  0.270000   23.000000    0.000000  
50%     33.200000                  0.452000   27.000000    0.000000  
75%     37.000000                  0.687000   36.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [4]:
pima.astype(float)
pima.corr()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            9.0    134.0           74.0           33.0     60.0  25.9   
1           10.0    101.0           76.0           48.0    180.0  32.9   
2            7.0    142.0           60.0           33.0    190.0  28.8   
3            6.0    129.0           90.0            7.0    326.0  19.6   
4            8.0    181.0           68.0           36.0    495.0  30.1   
5            1.0    189.0           60.0           23.0    846.0  30.1   
6            5.0    144.0           82.0           26.0    285.0  32.0   
7            8.0    176.0           90.0           34.0    300.0  33.7   
8           12.0    140.0           82.0           43.0    325.0  39.2   
9            0.0    173.0           78.0           32.0    265.0  46.5   
10          13.0    145.0           82.0           19.0    110.0  22.2   
11           8.0    196.0           76.0           29.0    280.0  37.5   
12           4.0    111.0           72.0           47.0    207.0  37.1   
13           7.0    195.0           70.0           33.0    145.0  25.1   
14           6.0    103.0           72.0           32.0    190.0  37.7   
15           7.0    150.0           78.0           29.0    126.0  35.2   
16           9.0    171.0          110.0           24.0    240.0  45.4   
17           2.0    197.0           70.0           45.0    543.0  30.5   
18           9.0    145.0           88.0           34.0    165.0  30.3   
19           5.0    187.0           76.0           27.0    207.0  43.6   
20           8.0    124.0           76.0           24.0    600.0  28.7   
21           3.0    176.0           86.0           27.0    156.0  33.3   
22          11.0    143.0           94.0           33.0    146.0  36.6   
23           7.0    181.0           84.0           21.0    192.0  35.9   
24           5.0    166.0           72.0           19.0    175.0  25.8   
25           7.0    136.0           74.0           26.0    135.0  26.0   
26          10.0    148.0           84.0           48.0    237.0  37.6   
27          11.0    155.0           76.0           28.0    150.0  33.3   
28           1.0    164.0           82.0           43.0     67.0  32.8   
29          11.0    138.0           74.0           26.0    144.0  36.1   
..           ...      ...            ...            ...      ...   ...   
359          0.0    139.0           62.0           17.0    210.0  22.1   
360          4.0     99.0           76.0           15.0     51.0  23.2   
361          1.0    143.0           74.0           22.0     61.0  26.2   
362          1.0    114.0           66.0           36.0    200.0  38.1   
363          1.0     96.0           64.0           27.0     87.0  33.2   
364          1.0     97.0           64.0           19.0     82.0  18.2   
365          2.0    117.0           90.0           19.0     71.0  25.2   
366          0.0     94.0           70.0           27.0    115.0  43.5   
367          0.0    134.0           58.0           20.0    291.0  26.4   
368          1.0    131.0           64.0           14.0    415.0  23.7   
369          1.0     99.0           72.0           30.0     18.0  38.6   
370          1.0     71.0           78.0           50.0     45.0  33.2   
371          0.0     78.0           88.0           29.0     40.0  36.9   
372          0.0    124.0           56.0           13.0    105.0  21.8   
373          2.0     99.0           60.0           17.0    160.0  36.6   
374          0.0    102.0           64.0           46.0     78.0  40.6   
375          0.0    126.0           86.0           27.0    120.0  27.4   
376          0.0    138.0           60.0           35.0    167.0  34.6   
377          1.0    139.0           62.0           41.0    480.0  40.7   
378          0.0     84.0           64.0           22.0     66.0  35.8   
379          1.0     89.0           24.0           19.0     25.0  27.8   
380          2.0     94.0           68.0           18.0     76.0  26.0   
381      

Pregnancies   Glucose  BloodPressure  SkinThickness  \
Pregnancies                  1.000000  0.197908       0.220154       0.086719   
Glucose                      0.197908  1.000000       0.211582       0.195021   
BloodPressure                0.220154  0.211582       1.000000       0.233444   
SkinThickness                0.086719  0.195021       0.233444       1.000000   
Insulin                      0.089581  0.585357       0.098292       0.184161   
BMI                         -0.050162  0.207216       0.304267       0.664068   
DiabetesPedigreeFunction     0.011728  0.138704      -0.015912       0.159177   
Age                          0.681890  0.343569       0.300034       0.165468   
Outcome                      0.230179  0.517662       0.191916       0.254739   

                           Insulin       BMI  DiabetesPedigreeFunction  \
Pregnancies               0.089581 -0.050162                  0.011728   
Glucose                   0.585357  0.207216                  0.138704   
BloodPressure             0.098292  0.304267                 -0.015912   
SkinThickness             0.184161  0.664068                  0.159177   
Insulin                   1.000000  0.228305                  0.136208   
BMI                       0.228305  1.000000                  0.159108   
DiabetesPedigreeFunction  0.136208  0.159108                  1.000000   
Age                       0.220654  0.062257                  0.086988   
Outcome                   0.305794  0.264821                  0.213510   

                               Age   Outcome  
Pregnancies               0.681890  0.230179  
Glucose                   0.343569  0.517662  
BloodPressure             0.300034  0.191916  
SkinThickness             0.165468  0.254739  
Insulin                   0.220654  0.305794  
BMI                       0.062257  0.264821  
DiabetesPedigreeFunction  0.086988  0.213510  
Age                       1.000000  0.340587  
Outcome                   0.340587  1.000000

# DecisionTreeClassifier

https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html
https://en.wikipedia.org/wiki/Decision_tree_learning#Metrics
https://medium.com/@mohtedibf/indepth-parameter-tuning-for-decision-tree-6753118a03c3


### If I make NaN to Mean(), I get LOWER accuracy 

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

# Create Decision Tree classifer object
clf = DecisionTreeClassifier(criterion='gini')

# Train Decision Tree Classifer
clf = clf.fit(X_train,y_train)

#Predict the response for test dataset
y_pred = clf.predict(X_test)

print("Accuracy:",accuracy_score(y_test, y_pred))

Accuracy: 0.7244897959183674


# KNeighborsClassifier
### if I make NaN to Mean(), i get:

    BEST: i: 27, Accuracy: 0.7878787878787878
### else is get:
    BEST: i: 4, Accuracy: 0.75


In [6]:
topI = 0
topY = 0
for i in range(1,40,1):
    clf = KNeighborsClassifier(i)
    clf = clf.fit(X_train,y_train)
    y_pred = clf.predict(X_test)
    score = accuracy_score(y_test, y_pred)
    if score > topY:
        topY=score
        topI = i
#     print("i: {0}, Accuracy: {1}".format(i,score))
print("BEST: i: {0}, Accuracy: {1}".format(topI,topY))

BEST: i: 7, Accuracy: 0.7959183673469388


# RandomForestClassifier
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html

In [7]:
clf = RandomForestClassifier(criterion='entropy',max_depth=5)

# Train Decision Tree Classifer
clf = clf.fit(X_train,y_train)

#Predict the response for test dataset
y_pred = clf.predict(X_test)

print("Accuracy:",accuracy_score(y_test, y_pred))

Accuracy: 0.7653061224489796


C:\Users\kali\Anaconda3\envs\idp3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


## Ensembling with VotingClassifier
https://stats.stackexchange.com/questions/190151/ensembling-with-votingclassifier
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.VotingClassifier.html#sklearn.ensemble.VotingClassifier

In [8]:
def benchmark_mustPrimeWithTrain_Test_BeforeHand():
    topI = 0
    topY = 0
    clf_KnC = KNeighborsClassifier(5)
    for i in range(1,40,1):
        clf = KNeighborsClassifier(i)
        clfF = clf.fit(X_train,y_train)
        y_pred = clfF.predict(X_test)
        score = accuracy_score(y_test, y_pred)
        if score > topY:
            topY=score
            topI = i
            clf_KnC=clfF

    #     print("i: {0}, Accuracy: {1}".format(i,score))
    print("BEST: i: {0}, Accuracy: {1}".format(topI,topY))


    alg1 = RandomForestClassifier(criterion='entropy',max_depth=5) #ExtraTreesClassifier()
    alg2 = GradientBoostingClassifier()
    alg3 = clf_KnC #KNeighborsClassifier(topI)

    eAlg = VotingClassifier(estimators=[('RandomForestClassifier',alg1),('GradientBoostingClassifier',alg2),('KNeighborsClassifier',alg3)],
                          voting='hard', weights=[1,1,1]).fit(X_train,y_train)

    #eAlg.fit(X_train,y_train)
    pred_ens = eAlg.predict(X_test)

    MSE= mean_squared_error(pred_ens,y_test)  
    MAD= mean_absolute_error(pred_ens,y_test)  
    ac = accuracy_score(pred_ens, y_test)

    print("MSE:", MSE)
    print("MAD:", MAD)
    print("AC:", ac)
    
benchmark_mustPrimeWithTrain_Test_BeforeHand()

BEST: i: 7, Accuracy: 0.7959183673469388
MSE: 0.19387755102040816
MAD: 0.19387755102040816
AC: 0.8061224489795918


C:\Users\kali\Anaconda3\envs\idp3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


# SVM.SVC
https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html

https://en.wikipedia.org/wiki/Confusion_matrix

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=1)
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)


#classifier = SVC(kernel = 'linear', random_state=0)
classifier = LinearSVC()

classifier.fit(X_train,y_train)
y_pred = classifier.predict(X_test)
ac = accuracy_score(y_test, y_pred)
ac

C:\Users\kali\Anaconda3\envs\idp3\lib\site-packages\sklearn\svm\base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
          intercept_scaling=1, loss='squared_hinge', max_iter=1000,
          multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
          verbose=0)

0.7863247863247863

# GaussianNB

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=1)
classifier = GaussianNB()
classifier.fit(X_train,y_train)
y_pred = classifier.predict(X_test)
ac = accuracy_score(y_test, y_pred)
ac

GaussianNB(priors=None, var_smoothing=1e-09)

0.8290598290598291

# lets discard 'bad' rows

In [11]:
col_names = ['pregnant', 'glucose', 'bp', 'skin', 'insulin', 'bmi', 'pedigree', 'age', 'label']

pima = pd.read_csv('https://raw.githubusercontent.com/alik604/Notebooks/master/Data%20Science-Datasets/pima%20indians%20diabetes%20dataset/diabetes-cleanData.csv' ,header=0,error_bad_lines=False)

# keeping our fetures from above 
feature_cols = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age']
X = pima[feature_cols] # Features
y = pima['Outcome'] # Target variable

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1) # 70% training and 30% test

pima.head(5)


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            9      134             74             33       60  25.9   
1           10      101             76             48      180  32.9   
2            7      142             60             33      190  28.8   
3            6      129             90              7      326  19.6   
4            8      181             68             36      495  30.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.460   81        0  
1                     0.171   63        0  
2                     0.687   61        0  
3                     0.582   60        0  
4                     0.615   60        1

In [12]:
pima.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   389.000000  389.000000     389.000000     389.000000  389.000000   
mean      3.208226  122.547558      70.642674      29.118252  156.210797   
std       3.041862   30.883067      12.538532      10.536701  119.249681   
min       0.000000   56.000000      24.000000       7.000000   14.000000   
25%       1.000000   99.000000      62.000000      21.000000   76.000000   
50%       2.000000  119.000000      70.000000      29.000000  126.000000   
75%       5.000000  143.000000      78.000000      37.000000  190.000000   
max      13.000000  198.000000     110.000000      63.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  389.000000                389.000000  389.000000  389.000000  
mean    33.046787                  0.523527   30.753213    0.326478  
std      7.038295                  0.345944   10.159084    0.469528  
min     18.200000                  0.085000   21.000000    0.000000  
25%     28.400000                  0.270000   23.000000    0.000000  
50%     33.200000                  0.452000   27.000000    0.000000  
75%     37.000000                  0.687000   36.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [13]:
benchmark_mustPrimeWithTrain_Test_BeforeHand()

BEST: i: 7, Accuracy: 0.7959183673469388
MSE: 0.21428571428571427
MAD: 0.21428571428571427
AC: 0.7857142857142857


C:\Users\kali\Anaconda3\envs\idp3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


# lets fix 'bad' rows

In [14]:
col_names = ['pregnant', 'glucose', 'bp', 'skin', 'insulin', 'bmi', 'pedigree', 'age', 'label']


pima = pd.read_csv('https://raw.githubusercontent.com/alik604/Notebooks/master/Data%20Science-Datasets/pima%20indians%20diabetes%20dataset/diabetes.csv' ,header=1, names=col_names,error_bad_lines=False)

# keeping our fetures from above 
feature_cols = ['pregnant','glucose', 'insulin', 'bmi', 'pedigree','age']
X = pima[feature_cols] # Features
y = pima.label # Target variable

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1) # 70% training and 30% test

#pima.head(10)
pima.describe()

pregnant     glucose          bp        skin     insulin         bmi  \
count  767.000000  767.000000  767.000000  767.000000  767.000000  767.000000   
mean     3.842243  120.859192   69.101695   20.517601   79.903520   31.990482   
std      3.370877   31.978468   19.368155   15.954059  115.283105    7.889091   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      1.000000   99.000000   62.000000    0.000000    0.000000   27.300000   
50%      3.000000  117.000000   72.000000   23.000000   32.000000   32.000000   
75%      6.000000  140.000000   80.000000   32.000000  127.500000   36.600000   
max     17.000000  199.000000  122.000000   99.000000  846.000000   67.100000   

         pedigree         age       label  
count  767.000000  767.000000  767.000000  
mean     0.471674   33.219035    0.348110  
std      0.331497   11.752296    0.476682  
min      0.078000   21.000000    0.000000  
25%      0.243500   24.000000    0.000000  
50%      0.371000   29.000000    0.000000  
75%      0.625000   41.000000    1.000000  
max      2.420000   81.000000    1.000000

In [15]:
pima= pima.astype(float)
pima.corr()

pregnant   glucose        bp      skin   insulin       bmi  \
pregnant  1.000000  0.128846  0.141197 -0.082495 -0.072999  0.017518   
glucose   0.128846  1.000000  0.152498  0.056381  0.332383  0.220955   
bp        0.141197  0.152498  1.000000  0.207308  0.089098  0.281777   
skin     -0.082495  0.056381  0.207308  1.000000  0.437974  0.392553   
insulin  -0.072999  0.332383  0.089098  0.437974  1.000000  0.198111   
bmi       0.017518  0.220955  0.281777  0.392553  0.198111  1.000000   
pedigree -0.033927  0.136903  0.041180  0.183498  0.185579  0.140546   
age       0.544018  0.262408  0.239571 -0.115873 -0.040942  0.035911   
label     0.221087  0.465856  0.064882  0.073265  0.131984  0.292695   

          pedigree       age     label  
pregnant -0.033927  0.544018  0.221087  
glucose   0.136903  0.262408  0.465856  
bp        0.041180  0.239571  0.064882  
skin      0.183498 -0.115873  0.073265  
insulin   0.185579 -0.040942  0.131984  
bmi       0.140546  0.035911  0.292695  
pedigree  1.000000  0.032738  0.173245  
age       0.032738  1.000000  0.236417  
label     0.173245  0.236417  1.000000

# should I change the Y_test part of the data? 

In [16]:
X_train= X_train.replace(0, np.nan) # make 0 into NaN
X_train=X_train.apply(lambda x: x.fillna(x.mean()),axis=0)
X_train.head(12)

pregnant  glucose     insulin   bmi  pedigree  age
118       4.0     99.0   51.000000  23.2     0.223   21
205       8.0    196.0  280.000000  37.5     0.605   57
506       1.0    130.0  170.000000  28.6     0.692   21
587       3.0    176.0  156.000000  33.3     1.154   52
34        4.0    103.0  192.000000  24.0     0.966   33
257       1.0    193.0  375.000000  25.9     0.655   24
216       6.0    125.0  120.000000  30.0     0.464   32
598       1.0    109.0  120.000000  23.1     0.407   26
760       9.0    170.0  156.221477  44.0     0.403   43
160       7.0    102.0  105.000000  37.2     0.204   45
713       3.0    102.0  156.221477  29.5     0.121   32
372       2.0    105.0   94.000000  34.9     0.225   25

In [17]:
benchmark_mustPrimeWithTrain_Test_BeforeHand()

BEST: i: 3, Accuracy: 0.7760416666666666
MSE: 0.21354166666666666
MAD: 0.21354166666666666
AC: 0.7864583333333334


C:\Users\kali\Anaconda3\envs\idp3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


## Now that the data is cleaned up, lets select the best features

In [18]:
pima = pd.read_csv('https://raw.githubusercontent.com/alik604/Notebooks/master/Data%20Science-Datasets/pima%20indians%20diabetes%20dataset/diabetes-cleanData.csv' ,header=0,error_bad_lines=False)
pima = pima.astype(float)

feature_cols = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age']

X = pima[feature_cols] # Features
y = pima['Outcome'] # Target variable

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1) # 70% training and 30% test
X_train = X_train.replace(0, np.nan).apply(lambda x: x.fillna(x.mean()),axis=0)
X_test = X_test.replace(0, np.nan).apply(lambda x: x.fillna(x.mean()),axis=0)

pima.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0          9.0    134.0           74.0           33.0     60.0  25.9   
1         10.0    101.0           76.0           48.0    180.0  32.9   
2          7.0    142.0           60.0           33.0    190.0  28.8   
3          6.0    129.0           90.0            7.0    326.0  19.6   
4          8.0    181.0           68.0           36.0    495.0  30.1   

   DiabetesPedigreeFunction   Age  Outcome  
0                     0.460  81.0      0.0  
1                     0.171  63.0      0.0  
2                     0.687  61.0      0.0  
3                     0.582  60.0      0.0  
4                     0.615  60.0      1.0

## Baseline: 

In [19]:
benchmark_mustPrimeWithTrain_Test_BeforeHand()

BEST: i: 7, Accuracy: 0.7959183673469388
MSE: 0.22448979591836735
MAD: 0.22448979591836735
AC: 0.7755102040816326


C:\Users\kali\Anaconda3\envs\idp3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


In [20]:
# Feature Importance with Extra Trees Classifier

def getFeatureImportance(X,y):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.99, random_state=1) # 70% training and 30% test

    model = ExtraTreesClassifier(n_estimators=10)#,random_state=1
    model.fit(X_train, y_train)
    return model.feature_importances_

featureImportances_base= getFeatureImportance(X,y)

for i in range(10): 
    featureImportances_base = featureImportances_base + getFeatureImportance(X,y)

#featureImportances_base   
l = dict(list(zip(X_train.columns, featureImportances_base)))

l.pop(min(l, key=l.get))
l.pop(min(l, key=l.get))

keepCol=list(l.keys())
keepCol

0.75

0.7999999999999999

['Insulin',
 'Glucose',
 'Age',
 'Pregnancies',
 'DiabetesPedigreeFunction',
 'SkinThickness']

In [21]:
pima = pd.read_csv('https://raw.githubusercontent.com/alik604/Notebooks/master/Data%20Science-Datasets/pima%20indians%20diabetes%20dataset/diabetes-cleanData.csv' ,header=0,error_bad_lines=False)
pima = pima.astype(float)

X = pima[np.sort(keepCol)] # Features UPDATED
y = pima['Outcome'] # Target variable

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1) # 70% training and 30% test
X_train = X_train.replace(0, np.nan).apply(lambda x: x.fillna(x.mean()),axis=0)
X_test = X_test.replace(0, np.nan).apply(lambda x: x.fillna(x.mean()),axis=0)

X_train.head()

Age  DiabetesPedigreeFunction  Glucose  Insulin  Pregnancies  \
230  25.0                     0.280    119.0     63.0     1.000000   
139  31.0                     1.224    128.0    182.0     2.000000   
251  25.0                     1.699     82.0    115.0     2.000000   
106  34.0                     0.323    130.0     79.0     3.000000   
246  25.0                     0.757    107.0     74.0     3.889796   

     SkinThickness  
230           47.0  
139           37.0  
251           22.0  
106           23.0  
246           30.0

In [22]:
benchmark_mustPrimeWithTrain_Test_BeforeHand()

BEST: i: 9, Accuracy: 0.7857142857142857
MSE: 0.1836734693877551
MAD: 0.1836734693877551
AC: 0.8163265306122449


C:\Users\kali\Anaconda3\envs\idp3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


# Lets try with Data cleaning + PCA

In [55]:
pima = pd.read_csv('https://raw.githubusercontent.com/alik604/Notebooks/master/Data%20Science-Datasets/pima%20indians%20diabetes%20dataset/diabetes-cleanData.csv' ,header=0,error_bad_lines=False)
pima = pima.astype(float)

feature_cols = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age']

X = pima[feature_cols] # Features
y = pima['Outcome'] # Target variable

#X = X.replace(0, np.nan).apply(lambda x: x.fillna(x.mean()),axis=0)

pima.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0          9.0    134.0           74.0           33.0     60.0  25.9   
1         10.0    101.0           76.0           48.0    180.0  32.9   
2          7.0    142.0           60.0           33.0    190.0  28.8   
3          6.0    129.0           90.0            7.0    326.0  19.6   
4          8.0    181.0           68.0           36.0    495.0  30.1   

   DiabetesPedigreeFunction   Age  Outcome  
0                     0.460  81.0      0.0  
1                     0.171  63.0      0.0  
2                     0.687  61.0      0.0  
3                     0.582  60.0      0.0  
4                     0.615  60.0      1.0

In [56]:
scalar = StandardScaler() 
# fitting 
scalar.fit(X) 
scaled_data = scalar.transform(X) 

# feature extraction
pca = PCA(n_components=3)
fit = pca.fit(scaled_data)
# summarize components
print("Explained Variance: %s" % fit.explained_variance_ratio_)
print(fit.components_)

StandardScaler(copy=True, with_mean=True, with_std=True)

Explained Variance: [0.3194998  0.19721597 0.14929103]
[[ 0.31380196  0.42605277  0.33116031  0.37974042  0.3704336   0.36149354
   0.16824132  0.41223011]
 [-0.55039344 -0.04958342 -0.02672038  0.41592194  0.05893726  0.51565072
   0.17753648 -0.4690576 ]
 [-0.2208238   0.47065363 -0.40559774 -0.29759509  0.57731855 -0.24398281
   0.25920749 -0.12898372]]


In [57]:
x_pca = pca.transform(scaled_data) 
df= pd.DataFrame(x_pca)
x_pca.shape 

(389, 3)

In [42]:
df.head(5)

0         1         2
0  2.329107 -3.848936 -1.369509
1  2.431910 -2.131920 -2.084000
2  1.714220 -2.151247  0.305476
3  1.144594 -3.656141  0.855905
4  3.615888 -2.038079  1.876007

In [58]:
X_train, X_test, y_train, y_test = train_test_split(x_pca, y, test_size=0.25, random_state=1)
benchmark_mustPrimeWithTrain_Test_BeforeHand()

BEST: i: 7, Accuracy: 0.826530612244898
MSE: 0.17346938775510204
MAD: 0.17346938775510204
AC: 0.826530612244898


C:\Users\kali\Anaconda3\envs\idp3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


### regarding my kmeans sub benchmark: 

with  X.replace(0, np.nan).apply(lambda x: x.fillna(x.mean()),axis=0)
   * Accuracy: 0.8367346938775511 
    
without 
   *  Accuracy: 0.826530612244898

# Datacleaning + Less Features + PCA 

In [44]:
pima = pd.read_csv('https://raw.githubusercontent.com/alik604/Notebooks/master/Data%20Science-Datasets/pima%20indians%20diabetes%20dataset/diabetes-cleanData.csv' ,header=0,error_bad_lines=False)
pima = pima.astype(float)

feature_cols = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age']

X = pima[np.sort(keepCol)] # Features
y = pima['Outcome'] # Target variable

X = X.replace(0, np.nan).apply(lambda x: x.fillna(x.mean()),axis=0)

scalar = StandardScaler().fit(X) 
scaled_data = scalar.transform(X) 

pca = PCA(n_components=3)
fit = pca.fit(scaled_data)

x_pca = pca.transform(scaled_data) 

In [45]:
X_train, X_test, y_train, y_test = train_test_split(x_pca, y, test_size=0.25, random_state=1) 
benchmark_mustPrimeWithTrain_Test_BeforeHand()

BEST: i: 9, Accuracy: 0.8163265306122449
MSE: 0.22448979591836735
MAD: 0.22448979591836735
AC: 0.7755102040816326


C:\Users\kali\Anaconda3\envs\idp3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


### so, Datacleaning + Less Features + PCA, is a bad idea, in the case


# TODO try with autoencoder for dim reduction 